broker/broker.go

```diff
@@ -145,61 +167,117 @@ func (b *Broker) SubmitWork(clientId string, msg *Message) {
 
 }
 
+func (b *Broker) Terminate() {
+	b.wpool.Terminate()
+}
+
 func (b *Broker) Start() {
 	if b == nil {
 		log.Error("broker is null")
-		return
 	}
+	b.StartServiceLoop()
+	b.StartCluster()
+}
 
-	if b.config.HTTPPort != "" {
-		go InitHTTPMoniter(b)
+func (b *Broker) StartServiceLoop() {
+	ok := false
+	tmpDelay := 10 * ACCEPT_MIN_SLEEP
+	for !ok {
+		ok = b.StartService()
+		if !ok {
+			log.Error("ran into service startup errors, now sleeping for %dms before retry",
+				zap.Duration("sleeping", tmpDelay/time.Millisecond))
+			time.Sleep(tmpDelay)
+			tmpDelay *= 2
+			if tmpDelay > ACCEPT_MAX_SLEEP {
+				tmpDelay = ACCEPT_MAX_SLEEP
+			}
+		}
+	}
+}
+
+func (b *Broker) StartService() bool {
+	ok := true
+	post := func(err error, closer io.Closer, worker func() error, name string) {
+		if err != nil {
+			log.Error("Cannot start " + name, zap.Error(err))
+			ok = false
+		} else {
+			b.services[name] = closer
+			go worker()
+		}
+	}
+	have := func(name string) bool {
+		_, ok := b.services[name]
+		return ok
+	}
+
+	if b.config.HTTPPort != "" && !have(HTTPService) {
+		err, closer, worker := InitHTTPMoniter(b)
+		post(err, closer, worker, HTTPService)
 	}
 
 	//listen client over tcp
-	if b.config.Port != "" {
-		go b.StartClientListening(false)
+	if b.config.Port != "" && !have(TCPService) {
+		err, closer, worker := b.StartClientListening(false)
+		post(err, closer, worker, TCPService)
 	}
 
 	//listen for cluster
-	if b.config.Cluster.Port != "" {
-		go b.StartClusterListening()
+	if b.config.Cluster.Port != "" && !have(ClusterService) {
+		err, closer, worker := b.StartClusterListening()
+		post(err, closer, worker, ClusterService)
 	}
 
 	//listen for websocket
-	if b.config.WsPort != "" {
-		go b.StartWebsocketListening()
+	if b.config.WsPort != "" && !have(WSService) {
+		err, closer, worker := b.StartWebsocketListening()
+		post(err, closer, worker, WSService)
 	}
 
 	//listen client over tls
-	if b.config.TlsPort != "" {
-		go b.StartClientListening(true)
+	if b.config.TlsPort != "" && !have(TLSService) {
+		err, closer, worker := b.StartClientListening(true)
+		post(err, closer, worker, TLSService)
 	}
+	return ok
+}
 
+func (b *Broker) StartCluster() {
 	//connect on other node in cluster
 	if b.config.Router != "" {
 		go b.processClusterInfo()
 		b.ConnectToDiscovery()
 	}
+}
 
+func (b *Broker) StopService() {
+	for _, closer := range b.services {
+		closer.Close()
+	}
+	b.services = make(map[string]io.Closer)
```

Comment:
I'm not quite sure what is best here: not only is it possible to `Close` the underlying listener, you can also `Close` the `Server` itself or call `Shutdown` on the server, which does a graceful shutdown, waiting for connections to finish (with a timeout provided by a passed `Context`). If we use `Shutdown`, `RegisterOnShutdown` would give a hook for closing websockets connections, which none of these approaches explicitly closes. https://pkg.go.dev/net/http#Server.RegisterOnShutdown

Reply:
If we want to have it super-clean:
 - stop listening first, so that no new connections come in
 - there is no server-side disconnect message, so I guess it is fine to just close the sending side of the TCP sockets/web sockets
 - then wait for EOF from the client (with a short timeout), and ignore any data on the connection till then
 - then close the server socket